File: 6_Transformation/transformation.py
import open3d as o3d
import numpy as np
import matplotlib.pyplot as plt
import copy
import os
import sys

print('Experiment 1:')
mesh = o3d.geometry.TriangleMesh.create_coordinate_frame()
# Translate in the x-direction and y-direction
mesh_tx = copy.deepcopy(mesh).translate((1.3, 0, 0))
mesh_ty = copy.deepcopy(mesh).translate((0, 1.3, 0))
print(f'Center of mesh: {mesh.get_center()}')
print(f'Center of mesh tx: {mesh_tx.get_center()}')
print(f'Center of mesh ty: {mesh_ty.get_center()}')
o3d.visualization.draw_geometries([mesh, mesh_tx, mesh_ty])


print('Experiment 2:')
# If [relative] is set to be 'False', the center of the geometry
# is translated directly to the position specified in the first argument.
mesh = o3d.geometry.TriangleMesh.create_coordinate_frame()
mesh_mv = copy.deepcopy(mesh).translate((2, 2, 2), relative=False)
print(f'Center of mesh: {mesh.get_center()}')
print(f'Center of translated mesh: {mesh_mv.get_center()}')
o3d.visualization.draw_geometries([mesh, mesh_mv])


print('Rotation Experiment:')
mesh = o3d.geometry.TriangleMesh.create_coordinate_frame()
mesh_r = copy.deepcopy(mesh)
R = mesh.get_rotation_matrix_from_xyz((np.pi / 2, 0, np.pi / 4))
mesh_r.rotate(R, center=(0, 0, 0))
o3d.visualization.draw_geometries([mesh, mesh_r])


print('Rotation with center False')
mesh = o3d.geometry.TriangleMesh.create_coordinate_frame()
mesh_r = copy.deepcopy(mesh).translate((2, 0, 0))
mesh_r.rotate(mesh.get_rotation_matrix_from_xyz((np.pi / 2, 0, np.pi / 4)),
              center=(0, 0, 0))
o3d.visualization.draw_geometries([mesh, mesh_r])


print('Scale Experiment:')
mesh = o3d.geometry.TriangleMesh.create_coordinate_frame()
mesh_s = copy.deepcopy(mesh).translate((2, 0, 0))
mesh_s.scale(0.5, center=mesh_s.get_center())
o3d.visualization.draw_geometries([mesh, mesh_s])


print('Scale with center False:')
mesh = o3d.geometry.TriangleMesh.create_coordinate_frame()
mesh_s = copy.deepcopy(mesh).translate((2, 1, 0))
mesh_s.scale(0.5, center=(0, 0, 0))
o3d.visualization.draw_geometries([mesh, mesh_s])

print('General Transformation')
mesh = o3d.geometry.TriangleMesh.create_coordinate_frame()
T = np.eye(4)
T[:3, :3] = mesh.get_rotation_matrix_from_xyz((0, np.pi / 3, np.pi / 2))
T[0, 3] = 1
T[1, 3] = 1.3
print(T)
mesh_t = copy.deepcopy(mesh).transform(T)
o3d.visualization.draw_geometries([mesh, mesh_t])

File: 6_Transformation/src/Transformation.cpp
#include <iostream>
#include <memory>
#include <thread>
#include <random>
#include <cstdlib>

#include <Eigen/Dense>

#include "open3d/Open3D.h"

#define PI 3.14159d

int main(int argc, char* argv[]) 
{
    //----- Translate -----//
    auto mesh = open3d::geometry::TriangleMesh::CreateCoordinateFrame();
    //open3d::visualization::DrawGeometries({mesh});
    auto mesh_tx = open3d::geometry::TriangleMesh::CreateCoordinateFrame();
    auto mesh_ty = open3d::geometry::TriangleMesh::CreateCoordinateFrame();
    mesh_tx -> Translate(Eigen::Vector3d(1.3, 0, 0));
    mesh_ty -> Translate(Eigen::Vector3d(0, 1.3, 0));
    std::cout << "mesh center position:" << mesh -> GetCenter() << std::endl;
    std::cout << "mesh_tx center position:" << mesh_tx -> GetCenter() << std::endl;
    std::cout << "mesh_tx center position:" << mesh_ty -> GetCenter() << std::endl;

    open3d::visualization::DrawGeometries({mesh, mesh_tx, mesh_ty}, "Translate");

    //----- Rotation -----//
    auto mesh1 = open3d::geometry::TriangleMesh::CreateCoordinateFrame();
    auto mesh1_r = open3d::geometry::TriangleMesh::CreateCoordinateFrame();
    auto R = mesh1_r -> GetRotationMatrixFromXYZ(Eigen::Vector3d(PI/2, 0, PI/4));
    mesh1_r ->Rotate(R, Eigen::Vector3d(0, 0, 0));
    open3d::visualization::DrawGeometries({mesh1, mesh1_r}, "Rotation");

    //----- Scale -----//
    auto mesh2 = open3d::geometry::TriangleMesh::CreateCoordinateFrame();
    auto mesh2_s = open3d::geometry::TriangleMesh::CreateCoordinateFrame(); 
    mesh2_s -> Translate(Eigen::Vector3d(3, 0, 0));
    mesh2_s -> Scale(0.5, Eigen::Vector3d(0, 0, 0));
    open3d::visualization::DrawGeometries({mesh2, mesh2_s}, "Scale");


    //----- General Transformation -----//
    auto mesh3 = open3d::geometry::TriangleMesh::CreateCoordinateFrame();
    auto mesh3_t = open3d::geometry::TriangleMesh::CreateCoordinateFrame();
    Eigen::Matrix<double, 4, 4> T;
    T = Eigen::MatrixXd::Identity(4,4);
    auto RR = mesh3 -> GetRotationMatrixFromXYZ(Eigen::Vector3d(0, PI/3, PI/2));
    for(int i = 0; i<3; i++){
        for(int j = 0; j <3; j++){
            T(i, j) = RR(i, j);
        }
    }
    T(0, 3) = 1;
    T(1,3) = 1.3;
    std::cout << "T: " << T << std::endl;
    mesh3_t ->Transform(T);
    open3d::visualization::DrawGeometries({mesh3, mesh3_t}, "General Transformation");

    return 0;
}